tests/test_services/test_scheduler.py

```diff
@@ -0,0 +1,152 @@
+import os
+import sys
+
+# Get the absolute path to the root directory of your project
+root_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
+
+# Add the root directory to the Python module search path
+sys.path.append(root_path)
+
+import time
+import datetime
+import pytest
+import tests
+
+from application.repository import (reminder as Reminder, user as User)
+from application.services import telegram, scheduler
+from dateutil.relativedelta import relativedelta
+
+sent_reminder = "placeholder"
+
+# Mock function for sending reminders
+def mock_send_reminder(user, message):
+    global sent_reminder
+    sent_reminder = f"Reminder for {user}: {message}"
+
+@pytest.fixture(scope="session")
+def setup_scheduler():
+    scheduler.start()
+
+    # Replace the send_reminder function with the mock
+    telegram.send_reminder = mock_send_reminder
+
+    user = User.add_user(
+        'Baburao',
+        '8881212',
+        'starGarage',
+        'riks'
+    )
+
+    reminder = Reminder.add_reminder(
+        user,
+        'Test Reminder',
+        datetime.datetime.now(),
+        'none',
+    )
+
+    yield reminder
+
+    scheduler.scheduler.remove_all_jobs()
+    scheduler.scheduler.shutdown()
+
+def test_reminder_sending(setup_scheduler):
+    global sent_reminder
+    reminder = setup_scheduler
+    now = datetime.datetime.now() + datetime.timedelta(seconds=2)
+
+    Reminder.update_description(reminder, "Uthale")
+    Reminder.update_datetime(reminder, now)
+
+    scheduler.schedule_reminder(reminder)
+    time.sleep(2)
+
+    assert sent_reminder == f"Reminder for Baburao: Test Reminder today at {now.time()}.\n\nUthale"
+
+def test_reminder_preloading(setup_scheduler):
+    reminder = setup_scheduler
+    now = datetime.datetime.now()
+
+    Reminder.update_datetime(reminder, now)
+    Reminder.update_recurrence(reminder, "daily")
+
+    job = scheduler.schedule_reminder(reminder)
+    assert job.name == "mock_send_reminder"
+
+    scheduler.scheduler.shutdown()
+    scheduler.start()
+
+    jobs = scheduler.scheduler.get_jobs()
+    assert len(jobs) == 1
+    assert jobs[0].name == "mock_send_reminder"
+
+    scheduler.scheduler.remove_all_jobs()
+
+def test_reminder_recurrence_daily(setup_scheduler):
```

Comment:
Maybe we should try to compile these into one with parameterized recurrences :D

Reply:
Done :D